### ISSUE TITLE
"Unable to set Cookie" with phantomjs on loadSessionSnapshot

### ISSUE THREAD

**Issue by osaris** [Codeception/Codeception#2900]
Hello,

I'm using Codeception (2.1.7) with phantomjs (2.1.1) and I'm using `loadSessionSnapshot` to avoid login on every test with the following code in my `AcceptanceTester` :

```
  public function login()
  {
    if ($this->loadSessionSnapshot('login'))
      return;
    $this->amOnPage('/en');
    $this->fillField('login', '***');
    $this->fillField('password', '***');
    $this->click('#btn-login');
    $this->saveSessionSnapshot('login');
  }
```

but when I have multiple tests with `$I->login();`, on second call, tests are crashing with :

`[Facebook\WebDriver\Exception\UnableToSetCookieException] {"errorMessage":"Unable to set Cookie"...`

I tried to move the `amOnPage` call of my `login` method to be sure that the browser is started but problem is the same...

Thanks for your help !

**Comment by sibben** [comment]
Hello Osaris,
Did you manage to solve this?  I got exactly the same error,  same login code also

**Comment by sibben** [comment]
I solved it like so


    public $loggedIn;
  	public function login()
	{
	     // if snapshot exists - skipping login
	     if ($this->loggedIn) return;
	     // logging in
	     $this->goToPage('/login');
	     $this->submitForm('#login', [
   			'form[_username]' => 'email@domain.com',
    		'form[_password]' => 'password'
		]);
	     // saving snapshot
	     $this->saveSessionSnapshot('login');
	     $this->loggedIn = true;
	}

**Comment by OnekO** [comment]
I have the same problem, everything is working ok with firefox/chrome, but phantomjs throws the 'Unable to set cookie' error. @sibben solutions is not working for me :sob: Any idea?

**Comment by sibben** [comment]
Sorry @OnekO ,  don't know why it won't work for you.

**Comment by thinkspill** [comment]
I'm having this same issue again when enabling Code Coverage. I had that same issue a few years back, which has now seemingly returned. This problem is so strange.

https://github.com/Codeception/Codeception/issues/845

**Comment by thinkspill** [comment]
Ok, I just found the cause of my particular Code Coverage collection problem.... 

Xdebug code coverage colleciton will crash when the 'opcache' is enabled.

Ensure the following is configured in your php.ini (CLI and web):

```
opcache.enable => Off => Off
opcache.enable_cli => Off => Off
```

**Comment by tomvo** [comment]
I'm also receiving this error but looking at the screenshots Selenium generates, the session has been loaded after all. I also use it in a login flow and it shows the logged-in version of the application in the screenshot.

**Comment by mfuehrer82** [comment]
+1 with codeception 2.1.8 and phantomJs 2.1.1

**Comment by jj-barlow** [comment]
I'm running into this issue currently, using the same versions as @mfuehrer82. 

Seems to be whenever the `Codeception\Module\WebDriver->setCookie()` func gets called within `src/Codeception/Module/WebDriver.php`, it falls over and throws the exception.

This is pretty dire as I don't believe there's another solution for running tests in this manner while also having JS compatibility.

**Comment by jmmerino** [comment]
Same error with codeception 2.1.8 and phantomjs 2.1.7

**Comment by gabrielef** [comment]
Same problem with phantomjs 2.1.1 and codeception 2.2.0-RC2

**Comment by JDGrimes** [comment]
According to [this](http://sqa.stackexchange.com/a/18838/18542), the problem may be fixed in the development version of phantomjs 2.2.0.

---

I thought that I was experiencing the same thing as @tomvo:

>I'm also receiving this error but looking at the screenshots Selenium generates, the session has been loaded after all. I also use it in a login flow and it shows the logged-in version of the application in the screenshot. So not sure why Webdriver throws this error.

However, I realized this was just an illusion. The screenshot shows the logged-in version of the web app, yes, but only because the first test logged in and the browser still has that old page loaded. It is not really logging in.

**Comment by lowerends** [comment]
The line that actually throws the error is this one:

```
WebDriverException::throwException($status, $message, $results);
```
which is located in `/vendor/facebook/webdriver/lib/Remote/HttpCommandExecutor.php:288`. Uncommenting that line seems to solve the issue, although I do get a "Bad gateway" after logging in the test user.

**Comment by Naktibalda** [comment]
@lowerends your solution simply hides the error, it doesn't fix the issue.

WebDriverException::throwException only throws exception if $status is not zero.

**Comment by lowerends** [comment]
Ok, so I've actually built phantomjs from source from the master branch (version 2.2.0-development, dated June 30 2016) and the issue remains (Codeception version 2.2.2).

I know commenting out the exception doesn't fix the issue, but it does lead to some (incomplete) code coverage being generated. 

Is there any workaround that can be used at this moment?

**Comment by thinkspill** [comment]
I had to revert to using PhpBrowser for any tests I needed to generate coverage for, that also didn't require using JavaScript.

I have a separate WebDriver-only suite for any tests that must interact with something dynamic that PhpBrowser can't handle, or that needs to do makeScreenShot kind of things. That suite is excluded from coverage collection, so the WebDriver error doesn't occur.

This has been my workaround after a lot of searching for a solution. 

A nice effect though is that PhpBrowser is faster than WebDriver, so helps to counteract the added slowness of coverage collection.

**Comment by lowerends** [comment]
I would definitely like to have coverage for the acceptance tests with phantomjs. 

Is this a problem that could be solved using Codeception, or is it something that needs to be fixed in WebDriver or phantomjs?

**Comment by rebajz** [comment]
i use chrome and phantom, and i have env DevChrome.yml , DevPhantom.yml, etc.

because loadSessionSnapshot not work in phantom, i edit my fuction...

```
public function automaticLoginAsSuperAdminPartnerInShop($ShopID)
    {
        $I = $this;
        $selectedEnv= $this->getScenario()->current('env');
        if (strpos(selectedEnv,"Phantom") === false) 
          {
           if ($I->loadSessionSnapshot('PrihlasenSuperAdminInPartnerInSomeShop')) return;
           }
        $I->loginAsSuperAdminPartner();
     }
```

**Comment by bgever** [comment]
Came across this thread while digging for the same exception, this PhantomJS issue is the root cause, and why it's also potentially fixed in the 2.2.0-dev: https://github.com/ariya/phantomjs/issues/14047

**Comment by IljaN** [comment]
I built phantomjs from master 

```bash
git clone git://github.com/ariya/phantomjs.git
cd phantomjs
git checkout master
git submodule init
git submodule update
python build.py
```

Still having the same Problem

**Comment by Emdaisy1** [comment]
Same problem with Codeception 2.2.5 and PhantomJS 2.2.1 
Guess the fix for 2.2.1 didn't happen?

**Comment by carltondickson** [comment]
Was trying PhantomJS 2.1.1 and Codeception 2.2.5.

Ended up just using PhantomJS 1.9.8 and my code that calls setCookie seems to work fine.

**Comment by boboldehampsink** [comment]
PhantomJS 1.9.8 works for me too

**Comment by jinzhencheng** [comment]
You should do like this:
`driver.get(...)`
then
`driver.add_cookie()`
...
However,it does work for me.

**Comment by sucitivel** [comment]
ditto @boboldehampsink @carltondickson

**Comment by drmmr763** [comment]
I had this issue and was not able to resolve it. Using PhantomJS 2.1. It is apparently a known issue that has been fixed, but the fix has not been officially released:

https://github.com/ariya/phantomjs/issues/14047

--

My solution was to switch away from PhantomJS to Selenium Server, using a docker file. I already had docker set up and running so my set up was very fast (15 minutes):

Per documentation on selenium headless testing with docker:
http://codeception.com/docs/modules/WebDriver#Headless-Selenium-in-Docker

I created a very simple docker-compose file:
```
version: "3"
services:
  web:
    image: selenium/standalone-chrome:3.4.0-chromium
    ports:
      - "4444:4444"
    network_mode: "host"
```

Modified my acceptance test yml:

```
 - WebDriver:
          url: http://xxx.yyy
          port: 4444
          browser: chrome
```

Running tests based on this is very simple now:

```
docker-compose up -d
php codecept run
docker-compose stop
```

Everything works with sessions and cookies as expected. Skipped the mess of configuring selenium, web drivers and installing various browsers.
